### PR TITLE
Add a new config-only option to prevent exclusive full screen mode

### LINKED ISSUE #7975: Running a Game in Fullscreen Disables the HDR Setting in Windows 10.
## Quick summary
Launching a game in fullscreen disables the HDR setting in Windows 10.

## Details
Here is the Windows 10 HDR setting:
[https://www.digitalcitizen.life/sites/default/files/gdrive/win10_hdr/hdr_10.png](https://www.digitalcitizen.life/sites/default/files/gdrive/win10_hdr/hdr_10.png
)

Launching a game with EPSC3 (0.0.9-10135-4ff69dc0 Alpha) in fullscreen mode or Alt-Enter into fullscreen will disable the HDR setting in the OS if it was already on. Not just an annoyance, but a serious problem with certain HDR displays that need to be power cycled any time this setting is changed or else color palette is totally out of whack. It happens with OpenGL or Vulcan. Using Nvidia GTX 1070. The two games I tested were BLES00276 and BCES00301.

### PR DESCRIPTION
**Pull request by 13xforever**
This is helpful for people streaming RPCS3, or to prevent disabling HDR mode in Windows.

Fixes #7975

New option is under `Vulkan` section of config.yml, it is disabled by default to maintain current behavior.
```yml
Video:
    Vulkan:
        Force Disable Exclusive Fullscreen Mode: false
```

People with multimonitor setup should test if switching to full screen works on different screens as expected.

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp [hunk 1]
```diff
@@ -517,6 +518,12 @@ namespace vk
 		return pgpu->external_memory_host_support;
 	}
 
+
```
Comment (elad335) [inline rpcs3/Emu/RSX/VK/vkutils/device.cpp:521]: wild whitespace.

File: rpcs3/Emu/RSX/VK/vkutils/device.cpp [hunk 2]
```diff
@@ -517,6 +518,11 @@ namespace vk
 		return pgpu->external_memory_host_support;
 	}
 
+	bool render_device::get_surface_capabilities_2_support() const
+	{
+		return pgpu->surface_capabilities_2_support;
+	}
+
```
Comment (MSuih) [inline rpcs3/Emu/RSX/VK/vkutils/device.cpp:525]: Whitespace on empty line.

File: rpcs3/Emu/system_config.h [hunk 3]
```diff
@@ -166,7 +166,8 @@ struct cfg_root : cfg::node
 			cfg::string adapter{ this, "Adapter" };
 			cfg::_bool force_fifo{ this, "Force FIFO present mode" };
 			cfg::_bool force_primitive_restart{ this, "Force primitive restart flag" };
-
+			cfg::_bool force_disable_exclusive_fullscreen_mode{this, "Force Disable Exclusive Fullscreen Mode"};
+
```
Comment (MSuih) [inline rpcs3/Emu/system_config.h:170]: Whitespace here as well.

File: rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp [hunk 4]
```diff
@@ -586,8 +588,49 @@ namespace vk
...
+			VkSurfaceCapabilities2KHR pSurfaceCapabilities = {};
+			pSurfaceCapabilities.sType                     = VK_STRUCTURE_TYPE_SURFACE_CAPABILITIES_2_KHR;
+
+			VkPhysicalDeviceSurfaceInfo2KHR pSurfaceInfo = {};
+			pSurfaceInfo.sType                           = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SURFACE_INFO_2_KHR;
+			pSurfaceInfo.surface                         = m_surface;
+
+			VkSurfaceCapabilitiesKHR surface_descriptors;
+			bool init_surface_descriptors = true;
+			bool should_disable_exclusive_full_screen = false;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp:600]: I think this can also be moved into the windows section

Comment (13xforever) [reply rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp:600]: `should_disable_exclusive_full_screen` yes, but not the `surface_descriptors` or the flag that it's still uninitialized

Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp:600]: Which is why I only marked this line 😹

File: rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp [hunk 5]
```diff
@@ -586,8 +586,49 @@ namespace vk
 			VkSwapchainKHR old_swapchain = m_vk_swapchain;
 			vk::physical_device& gpu = const_cast<vk::physical_device&>(dev.gpu());
 
-			VkSurfaceCapabilitiesKHR surface_descriptors = {};
-			CHECK_RESULT(vkGetPhysicalDeviceSurfaceCapabilitiesKHR(gpu, m_surface, &surface_descriptors));
+			VkSurfaceCapabilities2KHR pSurfaceCapabilities = {};
```
Comment (kd-11) [inline rpcs3/Emu/RSX/VK/vkutils/swapchain.hpp:589]: This method is uncomfortably long now, I'll have to refactor it later. Seems fine otherwise.

**Review by kd-11 — APPROVED** [review]

**Comment by Augusto7743** [comment]
That feature is for Linux too ?

**Comment by 13xforever** [comment]
Please do not necropost. This feature was superseded by more flexible option, and it's a vulkan feature. Whether drivers advertise or support it is another matter. In the future please use discord or support forum to ask questions.

**Pull request merged**